xwayland/xwm.c

```diff
@@ -1361,6 +1372,70 @@ static void xwm_handle_net_active_window_message(struct wlr_xwm *xwm,
 	wlr_signal_emit_safe(&surface->events.request_activate, surface);
 }
 
+static void pending_startup_id_destroy(struct pending_startup_id *pending) {
+	wl_list_remove(&pending->link);
+	free(pending->msg);
+	free(pending);
+}
+
+static void xwm_handle_net_startup_info_message(struct wlr_xwm *xwm,
+		xcb_client_message_event_t *ev) {
+	struct pending_startup_id *pending, *curr = NULL;
+	wl_list_for_each(pending, &xwm->pending_startup_ids, link) {
+		if (pending->window == ev->window) {
+			curr = pending;
+			break;
+		}
+	}
+
+	char *start;
+	if (curr) {
+		curr->msg = realloc(curr->msg, curr->len + STARTUP_INFO_BUF_LEN);
+		if (!curr->msg) {
+			pending_startup_id_destroy(curr);
+		}
+		start = curr->msg + curr->len;
+		curr->len += STARTUP_INFO_BUF_LEN;
+	} else {
+		curr = calloc(1, sizeof(struct pending_startup_id));
+		if (!curr)
+			return;
+		curr->window = ev->window;
+		curr->msg = malloc(STARTUP_INFO_BUF_LEN);
+		if (!curr->msg) {
+			free(curr);
+			return;
+		}
+		start = curr->msg;
+		curr->len = STARTUP_INFO_BUF_LEN;
+		wl_list_insert(&xwm->pending_startup_ids, &curr->link);
+	}
+
+	char *id = NULL;
+	const char *data = (const char *)ev->data.data8;
+	for (size_t i = 0; i < STARTUP_INFO_BUF_LEN; i++) {
+		start[i] = data[i];
+		if (start[i] == '\0') {
+			if (strncmp(curr->msg, STARTUP_INFO_REMOVE_PREFIX,
+					strlen (STARTUP_INFO_REMOVE_PREFIX)) == 0 &&
+					strlen (curr->msg) > strlen(STARTUP_INFO_REMOVE_PREFIX)) {
+				id = curr->msg + strlen(STARTUP_INFO_REMOVE_PREFIX);
+				break;
+			} else {
+				wlr_log(WLR_ERROR, "Unhandled message '%s'\n", curr->msg);
+				pending_startup_id_destroy(curr);
+				return;
+			}
+		}
+	}
+
+	if (id) {
+		wlr_log(WLR_DEBUG, "Got startup id: %s", id);
+		wlr_signal_emit_safe(&xwm->xwayland->events.remove_startup_info, id);
```

Comment:
We might want to provide more info to compositors here. Can we introduce `struct wlr_xwayland_remove_startup_info_event` with a `const char *id` field and a window field?

This will also allow us to extend the event in the future if need be.